### PR TITLE
Retrofit converter

### PR DESCRIPTION
**Pull request by nomisRev**

### CODE REVIEW

**Comment by nomisRev** [comment]
While playing with Retrofit I also wrote a `CallAdapter` for Arrow. This is inherently not compile safe. Thoughts? https://gist.github.com/nomisRev/539dc9f7c306571e0bf6405f440efc2d

File: settings.gradle [hunk 1]
```diff
@@ -7,5 +7,5 @@ include ':helios-parser'
 include ':helios-benchmarks'
 include ':helios-sample'
 include ':helios-optics'
-include 'helios-dsl-syntax-compiler'
-
+include ':helios-dsl-syntax-compiler'
+include ':helios-retrofit-converter'
```
Comment (raulraja) [inline settings.gradle]: I would name this just `helios-retrofit` so we have a consistent naming convention for integration modules.

**Comment by nomisRev** [comment]
@raulraja any thoughts on the `CallAdapter` for Arrow?

**Comment by raulraja** [comment]
@nomisRev looked at the code and originally looks good but I have some thoughts I'd like to discuss with you to see if we can make the impl less dynamic. Also is it common to accept all ApplicativeError instances? I'd have expected this adapter to be only valid for `Async` and up unless retrofit is always blocking being the scenes.

**Comment by nomisRev** [comment]
@raulraja it was just a small prototype I made to see what was possible. Sadly Retrofit heavily relies on reflection and tends to blow up @ runtime.

Concerning the Async / Sync most of the times both are offered AFAIK. [RxJava 2](https://github.com/square/retrofit/blob/687b3a639d64dfddd549d79a0097614dd1ef2040/retrofit-adapters/rxjava2/src/main/java/retrofit2/adapter/rxjava2/RxJava2CallAdapterFactory.java#L58) offers a `create` / `createAsync` and `createWithScheduler` factory method.

We could do the same by defining.

```
object ArrowCallAdapter {
    inline fun <reified F> create(AE: ApplicativeError<F, Throwable> = applicativeError()): CallAdapter.Factory =
            arrowCallAdapterFactory(AE)

    inline fun <reified F> createAsync(AE: Async<F> = async()): CallAdapter.Factory =
            TODO()
}
```

The reason I think both are really valuable is that people might be using `Try` for network stuff in coroutines and might need/want the extra power of `Async`.

As far as flexibility is concerned the implementation shown below are not allowed.
`fun <F> repo(@Path("repoId") repoId: String): HK<F, Json>` or
```
interface Service<F> { ... }
interface TryService : Service<TryHK>
```

**Comment by raulraja** [comment]
what prevents retrofit from understanding that? Do they not deal with generics at all?

**Comment by pakoito** [comment]
@raulraja Generics are not a thing on Android :D

**Comment by nomisRev** [comment]
No Retrofit cannot deal with generics at all.

In the first case `fun <F> repo(@Path("repoId") repoId: String): HK<F, Json>` Retrofit bails because it does not support generics in the return type.
>java.lang.IllegalArgumentException: Method return type must not include a type variable or wildcard: arrow.HK<F, helios.core.Json>

In the second case of `Service<F>` Retrofit bails because an API interface must not extend another interface.

> java.lang.IllegalArgumentException: API interfaces must not extend other interfaces.

**Pull request merged**